internal/speakerlist/speakerlist.go

```diff
@@ -97,193 +91,158 @@ func New(logger log.Logger, nodeName, bindAddr, bindPort, secret, namespace, lab
-	// TODO: The k8s client parameter should ideally be a parameter to
-	// New(). However, that is not possible today because:
-	//
-	// 1. The newController() function takes the sList as param[1]
-	// 2. The newController() function uses the sList param to create the layer 2 controller[2]
-	// 3. The new Kubernetes client uses the created controller to set the callbacks[3]
-	//
-	// Then, we have a dependency cycle if we add the Kubernetes client to speakerlist.New():
-	//
-	// 1. To create a Kubernetes client we need a controller
-	// 2. The newController() function uses the sList param to create the layer 2 controller[2]
-	// 3. The new Kubernetes client uses the created controller to set the callbacks[3]
-	//
-	// We should probably move the speakerlist code to be an implementation
-	// detail of the Layer 2 controller and, when doing so, try to fix this
-	// problem.
-	//
-	// [1]: https://github.com/metallb/metallb/pull/662/files#diff-60053ad6fecb5a3cfabb6f3d9e720899R109
-	// [2]: https://github.com/metallb/metallb/pull/662/files#diff-60053ad6fecb5a3cfabb6f3d9e720899L232-L251
-	// [3]: https://github.com/metallb/metallb/pull/662/files#diff-60053ad6fecb5a3cfabb6f3d9e720899L160-L162
-
-	sl.client = client
-
-	if sl.ml == nil {
-		return
-	}
-
-	// Initialize sl.mlSpeakerIPs.
-	iplist, err := sl.mlSpeakers()
-	if err != nil {
-		level.Error(sl.l).Log("op", "memberDiscovery", "error", err, "msg", "failed to get pod IPs")
-		iplist = nil
-	}
-
-	sl.mlMux.Lock()
-	sl.mlSpeakerIPs = iplist
-	sl.mlMux.Unlock()
-
-	// Update mlSpeakerIPs in the background.
-	go sl.updateSpeakerIPs()
-
+	sl.mlJoinCh = make(chan struct{}, 1)
+	sl.stopCh = make(chan struct{})
 	go sl.memberlistWatchEvents()
 	go sl.joinMembers()
+
+	return &sl, nil
 }
 
-// updateSpeakerIPs runs forever updating the sl.mlSpeakerIPs slice with the
-// IPs of all the speaker pods. As the function queries the API server, it waits at
-// least 5 minutes between queries to avoid self-induced API server DoS (should
-// be treated with care).
-func (sl *SpeakerList) updateSpeakerIPs() {
+func (sl *SpeakerList) joinMembers() {
 	for {
-		// This either stops the loop or makes sure to wait at least 5
-		// minutes before making another call to sl.mlSpeakers(), which
-		// queries the API server.
 		select {
+		case <-sl.mlJoinCh:
+			sl.Lock()
+			joinList := sl.mlJoinList
+			sl.Unlock()
+			if len(joinList) == 0 {
+				// joinList is empty, let's wait for a new rejoin() call
+				break
+			}
+			// Join is sequential and TCPTimeout = 10s, so it can take time.
+			n, err := sl.ml.Join(joinList)
+			if err == nil {
+				level.Info(sl.l).Log("op", "ml-join", "msg", "memberlist join", "nb joigned", n, "join list", joinList)
+			} else {
+				level.Error(sl.l).Log("op", "ml-join", "msg", "memberlist join", "nb joigned", n, "error", err, "join list", joinList)
+			}
+			select {
+			case <-time.After(15 * time.Second):
+				// Call rejoin() to maybe join again
```

Comment:
Looks like we want to explain *why* we are calling `rejoin()` (what does "maybe join again" mean?). The fact that we're calling `rejoin()` is obvious from the code.

Also - missing period.

Reply:
```
// Wait 15s between join attemps, by calling rejoin() we go to 'case <-sl.mlJoinCh',
// so if mlJoinList is now empty we will break.
```

Reply:
- `s/attemps/attempts/`
- What does "now" mean here?

Sorry, I am still confused by this comment. I don't understand its purpose. Does the comment explain why we wait? Does it explain why we're calling `sl.rejoin()`? Does it explain something else?

Reply:
```
// Wait 15s between join attempts, by calling rejoin() we go to 'case <-sl.mlJoinCh',
// so if mlJoinList is not empty after MemberList.Join() we will try joining again,
// and if it's empty we will break right away.
```